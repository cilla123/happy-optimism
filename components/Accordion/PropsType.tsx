export interface BaseAccordionProps  {
  activeIndex?: string | number | Array<string | number>;
  defaultActiveIndex?: string | number | Array<string | number>;
  multiple?: boolean;
  animated?: boolean;
  open?: boolean;
  onChange: (key: number) => void;
}

export interface BaseAccordionItemProps {
  title: string | JSX.Element;
  index: string;
  activeIndex?: Array<string>;
  animated?: boolean;
  open?: boolean;
  onItemChange?: (key: string) => void;
}
