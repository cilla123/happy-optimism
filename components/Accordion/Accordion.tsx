import React, { PureComponent, Children, cloneElement } from 'react';
import classnames from 'classnames';
import { BaseAccordionProps } from './PropsType';
import { isArray } from '../utils/validate';

export interface AccordionProps extends BaseAccordionProps {
  prefixCls?: string;
  className?: string;
}

export default class Accordion extends PureComponent<AccordionProps, any> {

  static defaultProps = {
    prefixCls: 'za-accordion',
    multiple: true,
    animated: false,
    open: false,
    onChange: () => {},
  };

  static Item: any;

  constructor(props) {
    super(props);

    this.state = {
      activeIndex: this.getActiveIndex(props),
    };
  }

  componentWillReceiveProps(nextProps) {
    if (!this.isPropEqual(this.props.activeIndex, nextProps.activeIndex)) {
      this.setState({
        activeIndex: this.getActiveIndex(nextProps),
      });
    }
  }

  onItemChange = (key) => {
    const { multiple, onChange } = this.props;
    const { activeIndex } = this.state;
    const hasKey = activeIndex.indexOf(key) > -1;

    let newActiveIndex: Array<string> = [];
    if (multiple) {
      if (hasKey) {
        newActiveIndex = activeIndex.filter(i => i !== key);
      } else {
        newActiveIndex = activeIndex.slice(0);
        newActiveIndex.push(key);
      }
    } else {
      newActiveIndex = hasKey ? [] : [key];
    }
    this.setState({
      activeIndex: newActiveIndex,
    });
    onChange(Number(key));
  }

  getActiveIndex(props) {
    const { activeIndex, defaultActiveIndex, multiple } = props;

    const defaultIndex = (activeIndex || activeIndex === 0) ? activeIndex : defaultActiveIndex;

    if (defaultIndex || defaultIndex === 0) {
      if (isArray(defaultIndex)) {
        return !multiple ?
        [String(defaultIndex[0])] : (defaultIndex as Array<any>).map(key => String(key));
      } else {
        return [String(defaultIndex)];
      }
    }

    return [];
  }

  isPropEqual(cur, next) {
    if (isArray(next) && isArray(cur)) {
     return next.length === cur.length && next.every((key, i) => key === cur[i]);
    }

    return cur === next;
  }

  renderItems() {
    const { animated, open } = this.props;
    const { activeIndex } = this.state;

    return Children.map(this.props.children, (ele, index) => {

      return cloneElement(ele as JSX.Element, {
        index: String(index),
        animated,
        activeIndex,
        open,
        onItemChange: this.onItemChange,
      });
    });
  }

  render() {
    const { prefixCls, className } = this.props;
    const cls = classnames(`${prefixCls}`, className);

    return (
      <div className={cls}>
        {this.renderItems()}
      </div>
    );
  }
}
