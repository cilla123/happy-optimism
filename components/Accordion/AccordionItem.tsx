import React, { PureComponent } from 'react';
import classnames from 'classnames';
import { BaseAccordionItemProps } from './PropsType';

export interface AccordionItemProps extends BaseAccordionItemProps {
  prefixCls?: string;
  className?: string;
}

export default class AccordionItem extends PureComponent<AccordionItemProps, any> {

  static defaultProps = {
    prefixCls: 'za-accordion',
  };

  private content;

  constructor(props) {
    super(props);

    this.state = {
      active: this.isActive(props),
    };
  }

  componentWillReceiveProps(nextProps) {
    const { active } = this.state;
    const { animated } = nextProps;

    if (active !== this.isActive(nextProps)) {
      this.setState({
        active: !active,
      });
      if (animated) {
        this.setStyle(active);
      }
    }
  }

  onClickItem = () => {
    const { index, onItemChange, animated, open } = this.props;
    const { active } = this.state;

    if (open) {
      return;
    }
    this.setState({
      active: !active,
    });
    if (animated) {
      this.setStyle(active);
    }
    if (onItemChange) {
      onItemChange(index);
    }
  }

  isActive(props) {
    const { index, activeIndex = [] } = props;
    return activeIndex.indexOf(index) > -1;
  }

  setStyle(active) {
    const { content } = this;
    let height;

    if (active) {
      height = content.offsetHeight;
      content.style.height = `${height}px`;

      setTimeout(() => {
        content.style.height = '0px';
      }, 0);
    } else {
      content.style.height = '0px';

      setTimeout(() => {
        content.style.height = `${this.getContentHeight(content)}px`;
      }, 0);
    }
  }

  getContentHeight(content) {
    const children = [...content.children];
    return children.reduce((res, next) => {
      res += next.offsetHeight;
      return res;
    }, 0);
  }

  getCls() {
    const { prefixCls, className, animated, open } = this.props;
    const { active } = this.state;

    const cls = classnames(`${prefixCls}-item`, className, {
      active: active || open,
    });
    const titleCls = `${prefixCls}-item-title`;
    const contentCls = classnames(`${prefixCls}-item-content`, {
      [`${prefixCls}-item-content-anim`]: animated,
    });
    const contentInnerCls = `${prefixCls}-item-content-inner`;
    const arrowCls = classnames(`${prefixCls}-item-arrow`, {
      [`${prefixCls}-item-arrow-hidden`]: open,
    });

    return { cls, titleCls, contentCls, contentInnerCls, arrowCls };
  }

  render() {
    const { title, children } = this.props;
    const { cls, titleCls, contentCls, contentInnerCls, arrowCls } = this.getCls();

    return (
      <div className={cls}>
        <div
          className={titleCls}
          onClick={this.onClickItem}
        >
          <div>{title}</div>
          <div className={arrowCls} />
        </div>
        <div
          className={contentCls}
          ref={(content) => this.content = content}
        >
          <div className={contentInnerCls}>
            {children}
          </div>
        </div>
      </div>
    );
  }
}
