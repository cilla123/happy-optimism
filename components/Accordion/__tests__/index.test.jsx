import React from 'react';
import { render, mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Accordion from '../index';

describe('Accordion', () => {
  const props = {};

  it('renders correctly', () => {
    const wrapper = render(
      <Accordion {...props}>
        <Accordion.Item title="50元套餐">
          <div>50元套餐内容</div>
        </Accordion.Item>
      </Accordion>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders correctly with accordion mode', () => {
    props.multiple = false;
    const wrapper = render(
      <Accordion {...props}>
        <Accordion.Item title="50元套餐">
          <div>50元套餐内容</div>
        </Accordion.Item>
      </Accordion>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders correctly with animated', () => {
    props.animated = true;
    const wrapper = render(
      <Accordion {...props}>
        <Accordion.Item title="50元套餐">
          <div>50元套餐内容</div>
        </Accordion.Item>
      </Accordion>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders correctly with defaultActiveIndex', () => {
    props.defaultActiveIndex = [1];
    const wrapper = render(
      <Accordion {...props}>
        <Accordion.Item title="50元套餐">
          <div>50元套餐内容</div>
        </Accordion.Item>
        <Accordion.Item title="100元套餐">
          <div>100元套餐内容</div>
        </Accordion.Item>
      </Accordion>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders correctly with string defaultActiveIndex', () => {
    props.defaultActiveIndex = '1';
    const wrapper = render(
      <Accordion {...props}>
        <Accordion.Item title="50元套餐">
          <div>50元套餐内容</div>
        </Accordion.Item>
        <Accordion.Item title="100元套餐">
          <div>100元套餐内容</div>
        </Accordion.Item>
      </Accordion>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders correctly with number defaultActiveIndex', () => {
    props.defaultActiveIndex = 0;
    const wrapper = render(
      <Accordion {...props}>
        <Accordion.Item title="50元套餐">
          <div>50元套餐内容</div>
        </Accordion.Item>
        <Accordion.Item title="100元套餐">
          <div>100元套餐内容</div>
        </Accordion.Item>
      </Accordion>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders correctly with activeIndex', () => {
    props.activeIndex = [0];
    const wrapper = render(
      <Accordion {...props}>
        <Accordion.Item title="50元套餐">
          <div>50元套餐内容</div>
        </Accordion.Item>
        <Accordion.Item title="100元套餐">
          <div>100元套餐内容</div>
        </Accordion.Item>
      </Accordion>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders correctly with string activeIndex', () => {
    props.activeIndex = '1';
    const wrapper = render(
      <Accordion {...props}>
        <Accordion.Item title="50元套餐">
          <div>50元套餐内容</div>
        </Accordion.Item>
        <Accordion.Item title="100元套餐">
          <div>100元套餐内容</div>
        </Accordion.Item>
      </Accordion>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders correctly with number activeIndex', () => {
    props.activeIndex = 0;
    const wrapper = render(
      <Accordion {...props}>
        <Accordion.Item title="50元套餐">
          <div>50元套餐内容</div>
        </Accordion.Item>
        <Accordion.Item title="100元套餐">
          <div>100元套餐内容</div>
        </Accordion.Item>
      </Accordion>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders correctly with dynamic activeIndex', () => {
    props.activeIndex = [0];
    const wrapper = shallow(
      <Accordion {...props}>
        <Accordion.Item title="50元套餐">
          <div>50元套餐内容</div>
        </Accordion.Item>
        <Accordion.Item title="100元套餐">
          <div>100元套餐内容</div>
        </Accordion.Item>
      </Accordion>
    );
    wrapper.setProps({ activeIndex: '1' });

    expect(wrapper.state('activeIndex')).toEqual(['1']);
  });

  it('click accordion item correctly', () => {
    props.onChange = jest.fn();
    const wrapper = mount(
      <Accordion {...props}>
        <Accordion.Item title="50元套餐" {...props}>
          <div>50元套餐内容</div>
        </Accordion.Item>
      </Accordion>
    );
    wrapper.find('.za-accordion-item-title').simulate('click');
    expect(props.onChange).toBeCalled();
  });

  it('renders correctly with open mode', () => {
    props.open = true;
    const wrapper = render(
      <Accordion {...props}>
        <Accordion.Item title="50元套餐">
          <div>50元套餐内容</div>
        </Accordion.Item>
        <Accordion.Item title="100元套餐">
          <div>100元套餐内容</div>
        </Accordion.Item>
      </Accordion>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('click accordion item correctly with open mode', () => {
    props.onChange = jest.fn();
    const wrapper = mount(
      <Accordion {...props}>
        <Accordion.Item title="50元套餐">
          <div>50元套餐内容</div>
        </Accordion.Item>
      </Accordion>
    );
    wrapper.find('.za-accordion-item-title').simulate('click');
    expect(props.onChange).not.toBeCalled();
  });
});
