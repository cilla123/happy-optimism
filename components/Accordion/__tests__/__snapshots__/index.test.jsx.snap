// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion renders correctly 1`] = `
<div
  class="za-accordion"
>
  <div
    class="za-accordion-item"
  >
    <div
      class="za-accordion-item-title"
    >
      <div>
        50元套餐
      </div>
      <div
        class="za-accordion-item-arrow"
      />
    </div>
    <div
      class="za-accordion-item-content"
    >
      <div
        class="za-accordion-item-content-inner"
      >
        <div>
          50元套餐内容
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Accordion renders correctly with accordion mode 1`] = `
<div
  class="za-accordion"
>
  <div
    class="za-accordion-item"
  >
    <div
      class="za-accordion-item-title"
    >
      <div>
        50元套餐
      </div>
      <div
        class="za-accordion-item-arrow"
      />
    </div>
    <div
      class="za-accordion-item-content"
    >
      <div
        class="za-accordion-item-content-inner"
      >
        <div>
          50元套餐内容
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Accordion renders correctly with activeIndex 1`] = `
<div
  class="za-accordion"
>
  <div
    class="za-accordion-item active"
  >
    <div
      class="za-accordion-item-title"
    >
      <div>
        50元套餐
      </div>
      <div
        class="za-accordion-item-arrow"
      />
    </div>
    <div
      class="za-accordion-item-content za-accordion-item-content-anim"
    >
      <div
        class="za-accordion-item-content-inner"
      >
        <div>
          50元套餐内容
        </div>
      </div>
    </div>
  </div>
  <div
    class="za-accordion-item"
  >
    <div
      class="za-accordion-item-title"
    >
      <div>
        100元套餐
      </div>
      <div
        class="za-accordion-item-arrow"
      />
    </div>
    <div
      class="za-accordion-item-content za-accordion-item-content-anim"
    >
      <div
        class="za-accordion-item-content-inner"
      >
        <div>
          100元套餐内容
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Accordion renders correctly with animated 1`] = `
<div
  class="za-accordion"
>
  <div
    class="za-accordion-item"
  >
    <div
      class="za-accordion-item-title"
    >
      <div>
        50元套餐
      </div>
      <div
        class="za-accordion-item-arrow"
      />
    </div>
    <div
      class="za-accordion-item-content za-accordion-item-content-anim"
    >
      <div
        class="za-accordion-item-content-inner"
      >
        <div>
          50元套餐内容
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Accordion renders correctly with defaultActiveIndex 1`] = `
<div
  class="za-accordion"
>
  <div
    class="za-accordion-item"
  >
    <div
      class="za-accordion-item-title"
    >
      <div>
        50元套餐
      </div>
      <div
        class="za-accordion-item-arrow"
      />
    </div>
    <div
      class="za-accordion-item-content za-accordion-item-content-anim"
    >
      <div
        class="za-accordion-item-content-inner"
      >
        <div>
          50元套餐内容
        </div>
      </div>
    </div>
  </div>
  <div
    class="za-accordion-item active"
  >
    <div
      class="za-accordion-item-title"
    >
      <div>
        100元套餐
      </div>
      <div
        class="za-accordion-item-arrow"
      />
    </div>
    <div
      class="za-accordion-item-content za-accordion-item-content-anim"
    >
      <div
        class="za-accordion-item-content-inner"
      >
        <div>
          100元套餐内容
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Accordion renders correctly with number activeIndex 1`] = `
<div
  class="za-accordion"
>
  <div
    class="za-accordion-item active"
  >
    <div
      class="za-accordion-item-title"
    >
      <div>
        50元套餐
      </div>
      <div
        class="za-accordion-item-arrow"
      />
    </div>
    <div
      class="za-accordion-item-content za-accordion-item-content-anim"
    >
      <div
        class="za-accordion-item-content-inner"
      >
        <div>
          50元套餐内容
        </div>
      </div>
    </div>
  </div>
  <div
    class="za-accordion-item"
  >
    <div
      class="za-accordion-item-title"
    >
      <div>
        100元套餐
      </div>
      <div
        class="za-accordion-item-arrow"
      />
    </div>
    <div
      class="za-accordion-item-content za-accordion-item-content-anim"
    >
      <div
        class="za-accordion-item-content-inner"
      >
        <div>
          100元套餐内容
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Accordion renders correctly with number defaultActiveIndex 1`] = `
<div
  class="za-accordion"
>
  <div
    class="za-accordion-item active"
  >
    <div
      class="za-accordion-item-title"
    >
      <div>
        50元套餐
      </div>
      <div
        class="za-accordion-item-arrow"
      />
    </div>
    <div
      class="za-accordion-item-content za-accordion-item-content-anim"
    >
      <div
        class="za-accordion-item-content-inner"
      >
        <div>
          50元套餐内容
        </div>
      </div>
    </div>
  </div>
  <div
    class="za-accordion-item"
  >
    <div
      class="za-accordion-item-title"
    >
      <div>
        100元套餐
      </div>
      <div
        class="za-accordion-item-arrow"
      />
    </div>
    <div
      class="za-accordion-item-content za-accordion-item-content-anim"
    >
      <div
        class="za-accordion-item-content-inner"
      >
        <div>
          100元套餐内容
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Accordion renders correctly with open mode 1`] = `
<div
  class="za-accordion"
>
  <div
    class="za-accordion-item active"
  >
    <div
      class="za-accordion-item-title"
    >
      <div>
        50元套餐
      </div>
      <div
        class="za-accordion-item-arrow za-accordion-item-arrow-hidden"
      />
    </div>
    <div
      class="za-accordion-item-content za-accordion-item-content-anim"
    >
      <div
        class="za-accordion-item-content-inner"
      >
        <div>
          50元套餐内容
        </div>
      </div>
    </div>
  </div>
  <div
    class="za-accordion-item active"
  >
    <div
      class="za-accordion-item-title"
    >
      <div>
        100元套餐
      </div>
      <div
        class="za-accordion-item-arrow za-accordion-item-arrow-hidden"
      />
    </div>
    <div
      class="za-accordion-item-content za-accordion-item-content-anim"
    >
      <div
        class="za-accordion-item-content-inner"
      >
        <div>
          100元套餐内容
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Accordion renders correctly with string activeIndex 1`] = `
<div
  class="za-accordion"
>
  <div
    class="za-accordion-item"
  >
    <div
      class="za-accordion-item-title"
    >
      <div>
        50元套餐
      </div>
      <div
        class="za-accordion-item-arrow"
      />
    </div>
    <div
      class="za-accordion-item-content za-accordion-item-content-anim"
    >
      <div
        class="za-accordion-item-content-inner"
      >
        <div>
          50元套餐内容
        </div>
      </div>
    </div>
  </div>
  <div
    class="za-accordion-item active"
  >
    <div
      class="za-accordion-item-title"
    >
      <div>
        100元套餐
      </div>
      <div
        class="za-accordion-item-arrow"
      />
    </div>
    <div
      class="za-accordion-item-content za-accordion-item-content-anim"
    >
      <div
        class="za-accordion-item-content-inner"
      >
        <div>
          100元套餐内容
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Accordion renders correctly with string defaultActiveIndex 1`] = `
<div
  class="za-accordion"
>
  <div
    class="za-accordion-item"
  >
    <div
      class="za-accordion-item-title"
    >
      <div>
        50元套餐
      </div>
      <div
        class="za-accordion-item-arrow"
      />
    </div>
    <div
      class="za-accordion-item-content za-accordion-item-content-anim"
    >
      <div
        class="za-accordion-item-content-inner"
      >
        <div>
          50元套餐内容
        </div>
      </div>
    </div>
  </div>
  <div
    class="za-accordion-item active"
  >
    <div
      class="za-accordion-item-title"
    >
      <div>
        100元套餐
      </div>
      <div
        class="za-accordion-item-arrow"
      />
    </div>
    <div
      class="za-accordion-item-content za-accordion-item-content-anim"
    >
      <div
        class="za-accordion-item-content-inner"
      >
        <div>
          100元套餐内容
        </div>
      </div>
    </div>
  </div>
</div>
`;
